arangod/IResearch/IResearchDocument.cpp

```diff
@@ -715,6 +705,272 @@ bool StoredValue::write(irs::data_output& out) const {
+}
+
+void InvertedIndexFieldIterator::next() {
+  TRI_ASSERT(valid());
+  if (_currentTypedAnalyzer) {
+    if (_currentTypedAnalyzer->next()) {
+      TRI_ASSERT(_primitiveTypeResetter);
+      TRI_ASSERT(_currentTypedAnalyzerValue);
+      TRI_ASSERT(_value._analyzer.get());
+      _primitiveTypeResetter(_value._analyzer.get(), _currentTypedAnalyzerValue->value);
+      return;
+    } else {
+      _currentTypedAnalyzer.reset();
+    }
+  }
+  _valueSlice = VPackSlice::noneSlice();
+  while (_begin != _end) {
+    while (!_arrayStack.empty() && _valueSlice.isNone()) {
+      if (_arrayStack.back().valid()) {
+        _valueSlice = *_arrayStack.back();
+        _arrayStack.back()++;
+      } else {
+        _arrayStack.pop_back();
+      }
+    }
+    if (_arrayStack.empty()) {
+      if (!_nameBuffer.empty()) {
+        if (++_begin == _end) {
+          TRI_ASSERT(!valid());
+          return; // exhausted
+        }
+        _nameBuffer.clear();
+      }
+      _valueSlice = get(_slice, _begin->first, arangodb::velocypack::Slice::noneSlice());
+    }
+    if (!_valueSlice.isNone() && (!_begin->first.back().shouldExpand || _valueSlice.isArray())) {
+      if (_nameBuffer.empty()) {
+        bool isFirst = true;
+        for (auto& a : _begin->first) {
+          if (!isFirst) {
+            _nameBuffer += NESTING_LEVEL_DELIMITER;
+          }
+          _nameBuffer.append(a.name);
+          isFirst = false;
+        }
+      }
+      TRI_ASSERT(_begin->second._pool);
+      switch (_valueSlice.type()) {
+        case VPackValueType::Null:
+          setNullValue();
+          return;
+        case VPackValueType::Bool:
+          setBoolValue(_valueSlice);
+          return;
+        case VPackValueType::Object:
+          if (setValue(_valueSlice, _begin->second)) {
+            return;
+          }
+          THROW_ARANGO_EXCEPTION_FORMAT(
+              TRI_ERROR_NOT_IMPLEMENTED,
+              "Inverted index does not support indexing objects and configured "
+              "analyzer does "
+              " not accept objects. Please use another analyzer to process an "
+              "object or exclude field '%s'"
+              " from index definition",
+              _nameBuffer.c_str());
+          break;
+        case VPackValueType::Array: {
+          if (_begin->first.back().shouldExpand) {
+            _arrayStack.push_back(VPackArrayIterator(_valueSlice));
+          } else if (setValue(_valueSlice, _begin->second)) {
+            return;
+          } else {
+            THROW_ARANGO_EXCEPTION_FORMAT(
+                TRI_ERROR_NOT_IMPLEMENTED,
+                "Configured analyzer does not accept arrays and field has no "
+                "expansion set. "
+                "Please use another analyzer to process an array or exclude "
+                "field '%s'"
+                " from index definition or enable expansion",
+                _nameBuffer.c_str());
+          }
+        } break;
```

Comment:
```suggestion
          if (_begin->first.back().shouldExpand) {
            _arrayStack.push_back(VPackArrayIterator(_valueSlice));
          } else if (!setValue(_valueSlice, _begin->second)) {
            THROW_ARANGO_EXCEPTION_FORMAT(
                TRI_ERROR_NOT_IMPLEMENTED,
                "Configured analyzer does not accept arrays and field has no "
                "expansion set. "
                "Please use another analyzer to process an array or exclude "
                "field '%s'"
                " from index definition or enable expansion",
                _nameBuffer.c_str());
          }
          return;
```

Reply:
We should continue the cycle here. Iterator added to the _arrayStack  should be immediately used.